### ISSUE TITLE
IMAPServer.__cbFetch can raise uncaught AttributeError

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1992]
|<img alt="tvachon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| tvachon reported|
|-|-|
|Trac ID|trac#1992|
|Type|defect|
|Created|2006-08-08 23:42:44Z|

`twisted.mail.imap4.IMAPServer.__cbFetch` will raise an `AttributeError` at line 1633 or 1634 in twisted/mail/imap4.py if:

   b. `self.blocked` is not `None` and
   a. `self._oldTimeout` has been deleted (with the `del self._oldTimeout` in a previous iteration of `__cbFetch`)

since the `StopIteration` exception will be raised whenever the iteration finishes (see below).
```
    def __cbFetch(self, results, tag, query, uid):
        if self.blocked is None:
            self.blocked = []
            self._oldTimeout = self.setTimeout(None)
        try:
            id, msg = results.next()
        except StopIteration:
            self.sendPositiveResponse(tag, 'FETCH completed')
            self._unblock()
            self.setTimeout(self._oldTimeout) # This statement can raise an Attribute Error
            del self._oldTimeout # And so can this one.
        else:
            self.spewMessage(id, msg, query, uid
                ).addCallback(lambda _: self.__cbFetch(results, tag, query, uid)
                ).addErrback(self.__ebSpewMessage
                )
```

This appears to be a rare case, but happens when large numbers of messages are added to my server via the Mac Mail client.

An easy fix to this is as follows:

```
    def __cbFetch(self, results, tag, query, uid):
        if self.blocked is None:
            self.blocked = []
            self._oldTimeout = self.setTimeout(None)
        try:
            id, msg = results.next()
        except StopIteration:
            self.sendPositiveResponse(tag, 'FETCH completed')
            self._unblock()
            self.setTimeout(getattr(self, "_oldTimeout", None)) # This takes care of this line
            try: # This try/except block should fix this line
                del self._oldTimeout
            except AttributeError:
                pass
        else:
            self.spewMessage(id, msg, query, uid
                ).addCallback(lambda _: self.__cbFetch(results, tag, query, uid)
                ).addErrback(self.__ebSpewMessage
                )
```

I'm attaching a patch for the fix, and will hopefully write unit tests within the next week (along with tests for tickets 1977 and 1978).

Attachments:

* [twisted-ticket-1992-20060808.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e3d/e3d9d95962c452f35e4ce7166b8d584f7b43adf0/752b1e197f4ddb02bc03c9c6e1fc9ad17774825f.diff) (653 bytes) - added by tvachon on 2006-08-09 00:01:44Z - Fix for this defect

<details><summary>Searchable metadata</summary>

```
trac-id__1992 1992
type__defect defect
reporter__tvachon tvachon
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1155080564000000 1155080564000000
changetime__1156290946000000 1156290946000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="tvachon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|tvachon commented|
|-|-|

Sorry, I was looking at an old version of imap4.py when I filed this. This defect appears on lines 1638 and 1639 in the latest subversion checkout.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to tvachon|
|-|-|

I suspect the attached patch will result in timeout support being lost at some point during the lifetime of a connection.

Here's an alternate proposal.

```python
Index: twisted/mail/imap4.py
===================================================================
--- twisted/mail/imap4.py       (revision 17811)
+++ twisted/mail/imap4.py       (working copy)
@@ -1634,9 +1634,9 @@
             id, msg = results.next()
         except StopIteration:
             self.sendPositiveResponse(tag, 'FETCH completed')
-            self._unblock()
             self.setTimeout(self._oldTimeout)
             del self._oldTimeout
+            self._unblock()
         else:
             self.spewMessage(id, msg, query, uid
                 ).addCallback(lambda _: self.__cbFetch(results, tag, query, uid)
```

I believe what is happening is this:

The client sends two fetch commands in rapid succession.  The first is parsed and dispatched to ```do_FETCH``` which results in ```__cbFetch``` which sets _oldTimeout and blocked before calling down to ```spewMessage``` which begins sending a response to the request.  This may take several reactor iterations to complete, due to the use of iterateInReactor to handle the potentially larger sections of the response.  Meanwhile the second fetch gets blocked and queued.  Eventually the ```spewMessage``` Deferred fires and ```__cbFetch``` is invoked again.  This time blocked is not None so the first suite is skipped.  ```results.next()``` raises a StopIteration and the except suite runs.  ```_unblock()``` runs which parses the 2nd queued fetch request, going down into ```do_FETCH``` and then ```__cbFetch```.  This time ```self.blocked``` is ```None``` so it is set to an empty list and _oldTimeout is set (to ```None```, incorrectly).  ```unblock``` returns and ```self.setTimeout(self._oldTimeout)``` runs, followed by ```del self._oldTimeout```.  At this point the timeout setting is correct but the remembered timeout value has been destroyed.  When the 2nd fetch request finishes and follows the ```StopIteration``` except suite, it triggers the ```AttributeError```.

The patch above should resolve this by sorting out the re-entrancy issues and avoiding corrupting the instance state.

Of course, a unit test for this case would be most helpful.

**Comment by twisted-trac** [comment]
|<img alt="tvachon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|tvachon commented|
|-|-|

Thanks for the comments, I figured my fix might be a little shortsighted. Your fix does appear to have fixed my bug, thanks!

I've been working on a unittest for this defect for about 6 hours now, and haven't made a whole lot of progress. The bug appears in the wild after a pretty long string of 

[tag] UID FETCH [number] BODY.PEEK[] 

commands, and happens fairly randomly. Do you have any suggestions on how this might be tested? 

Thanks very much!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

I'll have a go at writing the test for this, I think I know how to do it.

Thanks for confirming that this fixes the problem for you.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Ready for review in fetch-pipeline-1992

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_7"></a>|@radix set owner to @exarkun|
|-|-|

The test is pretty creepy, but is sufficiently documented.

The code in that StopIteration except suite is pretty dense; a comment documenting the ordering requirements would be good.

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [17925]) Merge fetch-pipeline-1992

Author: exarkun
Reviewer: radix
Fixes [#1992](https://github.com/twisted/twisted/issues/1992)

This fixes a state corruption bug in the IMAP4 server which manifest when
a client pipelined three or more FETCH commands which included a request
for message text.  The server is now careful to avoid re-entrancy issues
when handling pipelining.  A test for this case has also been added, and
the IMAP4 server's scheduler is now parameterized.

Thanks to tvachon for pointing out this bug.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
